homeassistant/components/system_bridge/coordinator.py

```diff
@@ -68,174 +69,139 @@ def __init__(
-                self.title,
-                exception,
-            )
-            self.last_update_success = False
-            self.async_update_listeners()
-        except TimeoutError as exception:
-            self.logger.warning(
-                "Timed out waiting for %s. Will retry: %s",
-                self.title,
-                exception,
-            )
-            self.last_update_success = False
-            self.async_update_listeners()
-
-        async def close_websocket(_) -> None:
-            """Close WebSocket connection."""
-            await self.websocket_client.close()
-
-        # Clean disconnect WebSocket on Home Assistant shutdown
-        self.unsub = self.hass.bus.async_listen_once(
-            EVENT_HOMEASSISTANT_STOP, close_websocket
-        )
+            await self.clean_disconnect()
 
     async def _async_update_data(self) -> SystemBridgeData:
         """Update System Bridge data from WebSocket."""
-        self.logger.debug(
-            "_async_update_data - WebSocket Connected: %s",
-            self.websocket_client.connected,
-        )
-        if not self.websocket_client.connected:
-            await self._setup_websocket()
+        if self.listen_task is None or not self.websocket_client.connected:
+            try:
+                await self.check_websocket_connected()
+
+                self.logger.debug("Create listener task for %s", self.title)
+                self.listen_task = self.hass.async_create_background_task(
+                    self._listen_for_data(),
+                    name="System Bridge WebSocket Listener",
+                    eager_start=False,
+                )
+                self.logger.debug("Listening for data from %s", self.title)
+
+                await self.websocket_client.register_data_listener(
+                    RegisterDataListener(modules=MODULES)
+                )
+                self.logger.debug("Registered data listener for %s", self.title)
+
+                self.last_update_success = True
+                self.async_update_listeners()
+            except AuthenticationException as exception:
+                self.logger.error(
+                    "Authentication failed at setup for %s: %s", self.title, exception
+                )
+                await self.clean_disconnect()
+                raise ConfigEntryAuthFailed from exception
+            except (ConnectionClosedException, ConnectionErrorException) as exception:
+                self.logger.warning(
+                    "Connection error occurred for %s: %s",
+                    self.title,
+                    exception,
+                )
+                await self.clean_disconnect()
+            except TimeoutError as exception:
+                self.logger.warning(
+                    "Timed out waiting for %s: %s",
+                    self.title,
+                    exception,
+                )
+                await self.clean_disconnect()
```

Comment:
Should we raise something here?

Reply:
Since we expect the device to disconnect (shutdown, sleep etc.) are there any "good" exceptions we can raise here? ConfigEntryNotReady comes to mind, but this would still fill up the error logs